// @base44/.vite-plugin-AmN16hKI/dist/injections/utils.js
/** Check if an element has instrumentation attributes */
export function isInstrumentedElement(element) {
    const htmlEl = element;
    return !!(htmlEl.dataset?.sourceLocation || htmlEl.dataset?.visualSelectorId);
}
/** Get the selector ID from an element's data attributes (prefers source-location) */
export function getElementSelectorId(element) {
    const htmlEl = element;
    return (htmlEl.dataset?.sourceLocation ||
        htmlEl.dataset?.visualSelectorId ||
        null);
}
export const ALLOWED_ATTRIBUTES = ["src"];
export const PLUGIN_ELEMENT_ATTR = "data-vite-plugin-element";
/** Find elements by ID - first try data-source-location, fallback to data-visual-selector-id */
export function findElementsById(id) {
    if (!id)
        return [];
    const sourceElements = Array.from(document.querySelectorAll(`[data-source-location="${id}"]`));
    if (sourceElements.length > 0) {
        return sourceElements;
    }
    return Array.from(document.querySelectorAll(`[data-visual-selector-id="${id}"]`));
}
/**
 * Update element classes by visual selector ID.
 * Uses setAttribute instead of className to support both HTML and SVG elements.
 */
export function updateElementClasses(elements, classes) {
    elements.forEach((element) => {
        element.setAttribute("class", classes);
    });
}
/** Set a single attribute on all provided elements. */
export function updateElementAttribute(elements, attribute, value) {
    if (!ALLOWED_ATTRIBUTES.includes(attribute)) {
        return;
    }
    elements.forEach((element) => {
        element.setAttribute(attribute, value);
    });
}
/** Collect attribute values from an element for a given allowlist. */
export function collectAllowedAttributes(element, allowedAttributes) {
    const attributes = {};
    for (const attr of allowedAttributes) {
        const val = element.getAttribute(attr);
        if (val !== null) {
            attributes[attr] = val;
        }
    }
    return attributes;
}
/**
 * Freeze all CSS animations and transitions on the page by injecting
 * scoped styles under `[data-visual-edit-active]` and programmatically
 * finishing (or pausing) every running animation.
 *
 * Plugin-owned elements (`[data-vite-plugin-element]`) are excluded so
 * the plugin UI stays animated.
 */
export function stopAnimations() {
    if (document.getElementById('freeze-animations'))
        return;
    document.documentElement.setAttribute('data-visual-edit-active', '');
    const animStyle = document.createElement('style');
    animStyle.id = 'freeze-animations';
    animStyle.textContent = `
    [data-visual-edit-active] *:not([${PLUGIN_ELEMENT_ATTR}]):not([${PLUGIN_ELEMENT_ATTR}] *),
    [data-visual-edit-active] *:not([${PLUGIN_ELEMENT_ATTR}]):not([${PLUGIN_ELEMENT_ATTR}] *)::before,
    [data-visual-edit-active] *:not([${PLUGIN_ELEMENT_ATTR}]):not([${PLUGIN_ELEMENT_ATTR}] *)::after {
      animation-play-state: paused !important;
      transition: none !important;
    }
  `;
    const pointerStyle = document.createElement('style');
    pointerStyle.id = 'freeze-pointer-events';
    pointerStyle.textContent = `
    [data-visual-edit-active] * { pointer-events: none !important; }
    [${PLUGIN_ELEMENT_ATTR}], [${PLUGIN_ELEMENT_ATTR}] * { pointer-events: auto !important; }
  `;
    const target = document.head || document.documentElement;
    target.appendChild(animStyle);
    target.appendChild(pointerStyle);
    document.getAnimations().forEach((a) => {
        // Skip animations on plugin UI elements
        const animTarget = a.effect?.target;
        if (animTarget instanceof Element && animTarget.closest(`[${PLUGIN_ELEMENT_ATTR}]`))
            return;
        try {
            a.finish(); // fast-forward to end state
        }
        catch {
            a.pause(); // finish() throws on infinite animations — pause instead
        }
    });
}
/**
 * Resume all previously frozen animations and remove the injected
 * freeze styles.  Cleans up the `data-visual-edit-active` attribute
 * from `<html>` so scoped selectors no longer match.
 */
export function resumeAnimations() {
    const animStyle = document.getElementById('freeze-animations');
    if (!animStyle)
        return;
    animStyle.remove();
    document.getElementById('freeze-pointer-events')?.remove();
    document.documentElement.removeAttribute('data-visual-edit-active');
    document.getAnimations().forEach((a) => {
        if (a.playState === 'paused') {
            try {
                a.play();
            }
            catch { /* animation target may have been removed */ }
        }
    });
}
/**
 * Hit-test the page at (`x`, `y`) and walk up the DOM to find the
 * nearest ancestor that carries instrumentation attributes
 * (`data-source-location` or `data-visual-selector-id`).
 *
 * Temporarily disables the pointer-events freeze sheet so the
 * browser's native `elementFromPoint` can reach the real target.
 */
export function findInstrumentedElement(x, y) {
    const pointerStyle = document.getElementById('freeze-pointer-events');
    if (pointerStyle)
        pointerStyle.disabled = true;
    const el = document.elementFromPoint(x, y);
    if (pointerStyle)
        pointerStyle.disabled = false;
    return el?.closest('[data-source-location], [data-visual-selector-id]') ?? null;
}
/**
 * Resolve which element should be hovered at (`x`, `y`), skipping the
 * currently selected element.  Returns the selector ID of the hovered
 * element, or `null` if the point is empty or hits the selected element.
 */
export function resolveHoverTarget(x, y, selectedElementId) {
    const element = findInstrumentedElement(x, y);
    if (!element)
        return null;
    const selectorId = getElementSelectorId(element);
    if (selectorId === selectedElementId)
        return null;
    return selectorId;
}
